core/src/docs/rfcs/3911_deleter_api.md

```diff
@@ -0,0 +1,170 @@
+- `deleter` is a new API that returns a `Deleter` instance.
+- `deleter_with` is an extension of the existing `deleter` API, which supports additional options, such as `recursive`.
+
+The following new options will be available for `delete_with` and `deleter_with`:
+
+- `recursive`: Enable recursive deletion.
+- `concurrent`: How many delete tasks can be performed concurrently?
+- `buffer`: How many files can be buffered for send in a single batch?
+
+Users can delete a file recursively in this way:
+
+```rust
+let _ = op.delete_with("path/to/file").recursive(true).await?;
+```
+
+Users can delete multiple files in this way:
+
+
+```rust
+let deleter = op.deleter().await?;
+
+// Add a single file to the deleter.
+deleter.delete(path).await?;
+
+// Add a stream of files to the deleter.
+deleter.delete_all(&mut lister).await?;
+
+// Close deleter, make sure all input files are deleted.
+deleter.close().await?;
+```
+
+`Deleter` also implements [`Sink`](https://docs.rs/futures/latest/futures/sink/trait.Sink.html), so all the methods of `Sink` are available for `Deleter`. For example, users can use [`forward`](https://docs.rs/futures/latest/futures/stream/trait.StreamExt.html#method.forward) to forward a stream of files to `Deleter`:
+
+```rust
+// Init a deleter to start batch delete tasks.
+let deleter = op.deleter().await?;
+// List all files that ends with tmp
+let lister = op.lister(path).await?
+  .filter(|x|future::ready(x.ends_with(".tmp")));
+
+// Forward all paths into deleter.
+lister.forward(deleter).await?;
+
+// Send all from a stream into deleter.
+deleter.send_all(&mut lister).await?;
+
+// Close the deleter.
+deleter.close().await?;
+```
+
+Users can control the behavior of `Deleter` by setting the options:
+
+```rust
+let deleter = op.deleter()
+  // Allow up to 8 concurrent delete tasks, default to 1.
+  .concurrent(8)
+  // Configure the buffer size to 1000, default value provided by services.
+  .buffer(1000)
+  .await?;
+
+// Add a single file to the deleter.
+deleter.delete(path).await?;
+
+// Add a stream of files to the deleter.
+deleter.delete_all(&mut lister).await?;
+
+// Close deleter, make sure all input files are deleted.
+deleter.close().await?;
+```
+
+In response to `Deleter` API, we will remove APIs like `remove`, `remove_via` and `remove_all`.
+
+- `remove` and `remove_via` could be replaced by `Deleter` directly.
+- `remove_all` could be replaced by `delete_with(path).recursive(true)`.
+
+# Reference-level explanation
+
+We will add a new associated type in `Accessor`:
+
+```rust
+trait Accessor {
+    ...
+    
+    type Deleter = oio::Delete;
+    type BlockingDeleter = oio::BlockingDelete;
+}
+```
+
+And the `delete` API will be changed to return a `Deleter` instead:
+
+```diff
+trait Accessor {
+-  async fn delete(&self, path: &str, args: OpDelete) -> Result<RpDelete>;
++  async fn delete(&self, path: &str, args: OpDelete) -> Result<(RpDelete, Self::Deleter)>;
```

Comment:
I can not understand why we changed our existing `delete` API. it seems not related to the current RFC.  Can you explain why we need to change it?

Reply:
> I can not understand why we changed our existing `delete` API. it seems not related to the current RFC. Can you explain why we need to change it?

Firstly, to enable the return of a `Deleter` to an `Operator`, the `Accessor` must provide a corresponding API. There are two potential solutions:

- Add a new `deleter` API
- Change existing `delete` API

However, the `deleter` API will overlap with the existing `delete` API. The latter is merely a specific instance that removes a single file or directory.

Reply:
So the RFC doc said, we introduced a new API `op.deleter()`, and it seems to have some conflict with this, Maybe we need to change one of the two?
```
let deleter = op.deleter().await?;

// Add a single file to the deleter.
deleter.delete(path).await?;

// Add a stream of files to the deleter.
deleter.delete_all(&mut lister).await?;

// Close deleter, make sure all input files are deleted.
deleter.close().await?;
```

Reply:
Why not keep the previous "delete" API? 
In other types of operations, there may also be some overlaps, and it would be better to provide a convenient way for the user.
``` rust
op.delete(&path).await?;
// Or prepare to do a lot of delete operations.
let deleter = op.deleter().await?;
```

Reply:
> So the RFC doc said, we introduced a new API `op.deleter()`, and it seems to have some conflict with this,

Seems you mixed public API and raw API here. 

In this RFC, we will introduce Public API like the following to users:

```rust
let _ = op.delete(path).await?;
// deleter is a sink, there are other apis upon it.
let deleter = op.deleter().await?;

// Also, we have `delete_with` to give users more powerful control.
let _ = op.delete_with(path).await?;
let deleter = op.delete_with().await?;
```

To implement this API, we will change the Raw API, mainly about the `Accessor`.

Reply:
Wow, that's enough explanation.

Reply:
> Do you want to add this explanation about updating the public API based on the `deleter` into this RFC as well?

I'm not sure why there's confusion. I made it very clear that:

> The following new API will be added to `Operator`:

Reply:
I added more words about them, PTAL.

Reply:
Thanks, @Xuanwo, I have mixed these APIs. Your explanation has addressed my confusion.